src/Compilers/CSharp/Portable/Symbols/ConstraintsHelper.cs

```diff
@@ -511,14 +521,168 @@ private static bool HasDuplicateInterfaces(NamedTypeSymbol type, ConsList<Symbol
                     set.Free();
                     return false;
             }
-            
+
             // very rare case. 
             // some implemented interfaces are related
             // will have to instantiate interfaces and check
             hasRelatedInterfaces:
             return type.InterfacesNoUseSiteDiagnostics(basesBeingResolved).HasDuplicates(TypeSymbol.EqualsIgnoringDynamicAndTupleNamesComparer);
         }
 
+        /// <summary>
+        /// Deep checking tuple names means checking the constraints on every constituent type and all their base types.
+        /// Use TupleNamesCheckKind.Shallow when the caller is already visiting constituent types.
+        /// </summary>
+        private static bool CheckTupleNamesConstraints(TypeSymbol type, ArrayBuilder<TypeParameterDiagnosticInfo> builder, TupleNamesCheckKind tupleNamesCheckKind)
+        {
+            // The skip list is used to avoid duplicate checks. It also avoid circular checks, as in `Derived : Base<Derived>`.
+            var skipList = PooledHashSet<TypeSymbol>.GetInstance();
+            bool result = CheckTupleNamesConstraints(type, builder, tupleNamesCheckKind, skipList);
+            skipList.Free();
+            return result;
+        }
+
+        private static bool CheckTupleNamesConstraints(TypeSymbol type, ArrayBuilder<TypeParameterDiagnosticInfo> builder, TupleNamesCheckKind tupleNamesCheckKind, HashSet<TypeSymbol> skipList)
+        {
+            // Note this only checks for tuple name errors introduced by referencing this type.
+            //
+            // So if the type does not include any tuples, its reference cannot introduce an error.
+            // But its declaration could. For instance:
+            //     `class Error : Base, I<(int notA, int notB)>`
+            //
+            // Similarly, if the base type doesn't use type arguments from the type, then the reference of this type cannot introduce an error.
+            // But the base type may already have one (reported elsewhere). For instance:
+            //     `class C : BaseCanBeSkipped<int>, ITest`
+            //     `class D : BaseNeedsChecking<T>, ITest`
+
+            Debug.Assert((object)type != null && (object)skipList != null);
+            if (skipList.Contains(type) || !type.ContainsTuple())
+            {
+                return true;
+            }
+
+            switch (tupleNamesCheckKind)
+            {
+                case TupleNamesCheckKind.Shallow:
```

Comment:
>TupleNamesCheckKind.Shallow [](start = 21, length = 27)

Is there still a reason to have TupleNamesCheckKind.Shallow and TupleNamesCheckKind.DeepExceptTopLevel modes? It doesn't feel like there should be a fundamental difference between them since we established that bases should always be checked using TupleNamesCheckKind.Deep mode.